File: src/main/java/com/multicampus/gamesungcoding/a11ymarketserver/feature/product/dto/ProductResponse.java
package com.multicampus.gamesungcoding.a11ymarketserver.feature.product.dto;

import com.multicampus.gamesungcoding.a11ymarketserver.feature.product.entity.Product;

import java.util.List;
import java.util.UUID;

public record ProductResponse(UUID productId,
                              String productName,
                              String productDescription,
                              String sellerName,
                              Boolean isA11yGuarantee,
                              Integer productPrice,
                              List<ProductImageResponse> productImages,
                              UUID parentCategoryId,
                              UUID categoryId,
                              String categoryName) {

    public static ProductResponse fromEntity(Product product) {

        var seller = product.getSeller();
        var category = product.getCategory();
        return new ProductResponse(
                product.getProductId(),
                product.getProductName(),
                product.getProductDescription(),
                seller.getSellerName(),
                seller.getIsA11yGuarantee(),
                product.getProductPrice(),
                product.getProductImages()
                        .stream()
                        .map(ProductImageResponse::fromEntity)
                        .toList(),
                category.getParentCategory().getCategoryId(),
                category.getCategoryId(),
                category.getCategoryName()
        );
    }
}

File: src/main/kotlin/com/multicampus/gamesungcoding/a11ymarketserver/feature/product/dto/ProductResponse.kt
package com.multicampus.gamesungcoding.a11ymarketserver.feature.product.dto

import com.multicampus.gamesungcoding.a11ymarketserver.feature.product.entity.Product
import java.util.*

data class ProductResponse(
    val productId: UUID,
    val productName: String,
    val productDescription: String,
    val sellerName: String,
    val isA11yGuarantee: Boolean,
    val productPrice: Int,
    val productImages: List<ProductImageResponse?>,
    val parentCategoryId: UUID,
    val categoryId: UUID,
    val categoryName: String
) {
    // TODO: Seller 변환시 삭제
    companion object {
        @JvmStatic
        fun fromEntity(product: Product): ProductResponse {
            val seller = product.seller
            val category = product.category
            return ProductResponse(
                product.productId,
                product.productName,
                product.productDescription,
                seller.sellerName,
                seller.isA11yGuarantee,
                product.productPrice,
                product.productImages.map { ProductImageResponse.fromEntity(it) },
                category.parentCategory.categoryId,
                category.categoryId,
                category.categoryName
            )
        }
    }
}
